### PR TITLE
Upgrade gradle to 7.0.

### PR DESCRIPTION
**Pull request by zach-klippenstein**

### CODE REVIEW

**Comment by zach-klippenstein** [comment]
Seems like we finally need to get off our super-old Dokka version to unblock this.

**Comment by zach-klippenstein** [comment]
Dokka upgrade in #421 should unblock this.

**Comment by zach-klippenstein** [comment]
Looks like we need to upgrade the ktlint plugin as well, latest version is 10.0.0.

**Comment by zach-klippenstein** [comment]
And detekt, 1.16.0.

**Comment by rjrjr** [comment]
Completely. I don't care even a little about detekt.

**Comment by zach-klippenstein** [comment]
Removed detekt in #426, rebasing this on top of that…

**Review by patrickyin — APPROVED** [review]

**Comment by zach-klippenstein** [comment]
Updated jmh (benchmarking) and ktlint gradle plugins to their latest versions, required for Gradle 7.

**Comment by zach-klippenstein** [comment]
Added a separate commit that updates gradle and AGP for the compose project. I did both there because AGP was so far behind.

File: compose/build.gradle.kts
Comment (zach-klippenstein) [inline compose/build.gradle.kts:110]: I just copied this in from the main project's build.gradle.

File: compose/build.gradle.kts
Comment (maybeYouShouldBeTheLawyer) [inline compose/build.gradle.kts:96]: A heads-up just in case you're interested in a follow-on PR: it looks like the import-ordering issue [has been (mostly) fixed](https://youtrack.jetbrains.com/issue/KTIJ-4813#focus=Comments-27-4205728.0-0) thanks to the introduction of a [layout table for Kotlin imports](https://github.com/JetBrains/kotlin/pull/3336). (But note the misbehavior cited at the end of that conversation, including a comment by the tried-and-true Valeriy Ovechkin.)

File: compose/build.gradle.kts
Comment (maybeYouShouldBeTheLawyer) [reply compose/build.gradle.kts:110]: Seems we _should_ be able to re-enable `indent` and `parameter-list-wrapping` (assuming we're on a recent-enough version of ktlint and there aren't any new Kotlin-compatibility issues) 🤞. 

963 was allegedly fixed [here](https://github.com/pinterest/ktlint/pull/961); 964 [here](https://github.com/pinterest/ktlint/pull/1022) (transitively via 918); 965 [here](https://github.com/pinterest/ktlint/pull/972).

**Review by maybeYouShouldBeTheLawyer — APPROVED** [review]
A couple comments about the disabled ktlint rules. Otherwise lgtm.

File: compose/build.gradle.kts
Comment (zach-klippenstein) [reply compose/build.gradle.kts:96]: I'm not interesting in diverging these effectively duplicate configs. We can follow up and address in the main project once these have merged.

**Pull request merged**
